src/parser.ts

```diff
@@ -13,6 +13,21 @@ let PlaceHolders = {
     ArrowParameterPlaceHolder: 'ArrowParameterPlaceHolder'
 };
 
+function upgradeToken(token) {
+    if (token && state.strict && token.type === Token.Identifier) {
+        if (scanner.isStrictModeReservedWord(token.value)) {
+            token.type = Token.Keyword;
+        }
+    }
+    return token;
+}
+
+function nextToken() {
+    const token = scanner.lex();
+    scanner.lookahead = upgradeToken(scanner.lookahead);
```

Comment:
Even here, the scanner is asking the parser to upgrade the token, which is produced by the scanner, and uses another method of the scanner to make that determination.